src/main/java/com/iota/iri/service/garbagecollector/MilestonePrunerJob.java

```diff
@@ -0,0 +1,302 @@
+
+/**
+ * This class represents a cleanup job for the {@link GarbageCollector}.
+ *
+ * It removes milestones and all of their directly and indirectly referenced transactions ( and the orphaned subtangles
+ * branching off of the deleted transactions). Therefore it is used by the
+ * {@link com.iota.iri.service.snapshot.SnapshotManager} to clean up milestones prior to a snapshot.
+ *
+ * It gets processed one milestone at a time persisting the progress after each step.
+ */
+public class MilestonePrunerJob extends GarbageCollectorJob {
+    /**
+     * Holds the milestone index where this job starts cleaning up (read only).
+     */
+    private final int startingIndex;
+
+    /**
+     * Holds the milestone index where this job stops cleaning up (read only).
+     */
+    private int targetIndex;
+
+    /**
+     * Holds the milestone index of the oldest milestone that was cleaned up already (the current progress).
+     */
+    private int currentIndex;
+
+    /**
+     * This method registers this job type in a {@link GarbageCollector}.
+     *
+     * It registers the {@link JobParser}, the {@link QueueProcessor} and the {@link QueueConsolidator} which will
+     * allow us to process jobs of this particular type.
+     *
+     * @param garbageCollector {@link GarbageCollector} that shall be able to process {@link MilestonePrunerJob}s
+     */
+    static void registerInGarbageCollector(GarbageCollector garbageCollector) {
+        garbageCollector.registerParser(MilestonePrunerJob.class, MilestonePrunerJob::parse);
+        garbageCollector.registerQueueProcessor(MilestonePrunerJob.class, MilestonePrunerJob::processQueue);
+        garbageCollector.registerQueueConsolidator(MilestonePrunerJob.class, MilestonePrunerJob::consolidateQueue);
+    }
+
+    /**
+     * This method consolidates the cleanup jobs by merging two or more jobs together if they are either both done or
+     * pending.
+     *
+     * It is used to clean up the queue and the corresponding garbage collector file, so it always has a size of less
+     * than 4 jobs.
+     *
+     * Since the jobs are getting processed from the beginning of the queue, we first check if the first two jobs are
+     * "done" and merge them into a single one that reflects the "done" status of both jobs. Consecutively we check if
+     * there are two or more pending jobs at the end that can also be consolidated.
+     *
+     * It is important to note that the jobs always clean from their startingPosition to the startingPosition of the
+     * previous job (or the {@code milestoneStartIndex} of the last global snapshot if there is no previous one) without
+     * any gaps in between which is required to be able to merge them.
+     *
+     * @throws GarbageCollectorException if an error occurs while persisting the state
+     */
+    private static void consolidateQueue(GarbageCollector garbageCollector, ArrayDeque<GarbageCollectorJob> jobQueue) throws GarbageCollectorException {
```

Comment:
because then its not generic anymore - we will add further job types that take care of cleaning up the floating transactions for example and they might use a consolidator